### PR TITLE
genearate image from volume while creating template

### PR DESCRIPTION
**Pull request by shangma**
harvester issue #571

### CODE REVIEW

**Comment by shangma** [comment]
@gitlawr @guangbochen  I understand that for issue #568, when we export volume to an image, we made the ui to ask for a namespace as an argument so that we can export the image to a different ns, but here one vm may have multiple pvc. So instead of asking a namespace argument for each of the pvc, I use the namespace of the vm itself. Let me know if this is appropriate. The code also assume that the pvc owned by a vm will share the same namespace. Let me know if this makes sense.

**Comment by gitlawr** [comment]
Make sense. The VM and PVCs should be in the same namespace

**Comment by shangma** [comment]
Add "To Test" section in the original [issue](https://github.com/harvester/harvester/issues/571#issuecomment-948086400).

File: pkg/apis/harvesterhci.io/v1beta1/template.go [hunk 1]
```diff
@@ -68,7 +68,7 @@ type VirtualMachineTemplateVersionSpec struct {
 	Description string `json:"description,omitempty"`
 
 	// +optional
-	ImageID string `json:"imageId,omitempty"`
+	ImageID []string `json:"imageId,omitempty"`
```
Comment (gitlawr) [inline pkg/apis/harvesterhci.io/v1beta1/template.go:71]: Any impact on upgrade with this API change?

File: pkg/api/vm/handler.go [hunk 2]
```diff
@@ -428,6 +454,7 @@ func (h *vmActionHandler) createTemplate(namespace, name string, input CreateTem
 				Description: fmt.Sprintf("Template drived from virtual machine [%s]", vmID),
 				VM:          removeMacAddresses(vm),
 				KeyPairIDs:  keyPairIDs,
+				ImageID:     imageIDs,
```
Comment (gitlawr) [inline pkg/api/vm/handler.go:457]: When we create a VM from the template, how do we know the mapping of the images and the volumes

Comment (shangma) [reply pkg/api/vm/handler.go:457]: @gitlawr  how about we use a map with a format such as this {"disk-0": "generated-image-id-0"; "disk-1": "generated-image-id-1"} ?

File: pkg/apis/harvesterhci.io/v1beta1/template.go (hunk 1 above)
Comment (shangma) [reply pkg/apis/harvesterhci.io/v1beta1/template.go:71]: @gitlawr  I did need to rebuild the webhook image to allow this changes. Other that, the change was functional, see screenshot [here](https://github.com/harvester/harvester/issues/571#issuecomment-948086400).

Comment (gitlawr) [reply pkg/apis/harvesterhci.io/v1beta1/template.go:71]: I mean how do we handle legacy resources that use string for ImageID in the previous version

File: pkg/api/vm/handler.go (hunk 2 above)
Comment (gitlawr) [reply pkg/api/vm/handler.go:457]: I think we can set the volumeClaimTempaltes annotation: https://github.com/harvester/harvester/blob/d05a6cc730feccd7e242798b8b23973d709ce7e5/pkg/data/template.go#L121

**Comment by guangbochen** [comment]
Close as this is resolved in another PR https://github.com/harvester/harvester/pull/3097; thanks for the contribution.

**Pull request closed without merging**